smart-cards/src/main/java/gov/va/api/health/smartcards/patient/PatientController.java

```diff
@@ -95,28 +130,72 @@ void initDirectFieldAccess(DataBinder dataBinder) {
       @Valid @RequestBody Parameters parameters,
       @RequestHeader(name = "Authorization", required = false) String authorization) {
     checkState(!StringUtils.isEmpty(id), "id is required");
-    var credentialTypes = validateCredentialType(parameters);
+    var credentialTypes = getCredentialTypes(parameters);
+    validateCredentialTypes(credentialTypes);
     Patient.Bundle patients = fhirClient.patientBundle(id, authorization);
     Patient patient = getPatientFromBundle(patients, id);
     Immunization.Bundle immunizations = fhirClient.immunizationBundle(patient, authorization);
+    lookupLocations(immunizations, authorization);
     List<MixedEntry> resources = new ArrayList<>();
     consumeBundle(patients, resources, x -> true, this::transform);
     consumeBundle(immunizations, resources, this::filter, this::transform);
+    // Index unique URLs and replace with 'resource:X' scheme
+    List<String> urls = indexAndReplaceUrls(resources);
     MixedBundle bundle = toBundle(resources);
     var vc = vc(bundle, credentialTypes);
-    var parametersResponse = parameters(vc);
+    var parametersResponse = parameters(vc, urls);
     return ResponseEntity.ok(parametersResponse);
   }
 
+  private void lookupLocations(Immunization.Bundle immunizations, String authorization) {
+    // keep track of locations we already looked up
+    Map<String, Bundle> locations = new HashMap<>();
+    for (Immunization.Entry entry : immunizations.entry()) {
+      Immunization imm = entry.resource();
+      String locationResourceId = Controllers.resourceId(imm.location().reference());
+      Location.Bundle locationBundle = locations.get(locationResourceId);
+      if (locationBundle == null) {
+        locationBundle = fhirClient.locationBundle(locationResourceId, authorization);
+        locations.put(locationResourceId, locationBundle);
+      }
+      Optional<Location.Entry> maybeEntry = locationBundle.entry().stream().findFirst();
+      maybeEntry.ifPresent(value -> imm.contained(List.of(value.resource())));
```

Comment:
So although DQ doesn't use contained resources, I think we still need to take care here to add to the existing list, not overwrite it.